### PR TITLE
Dev

### LINKED ISSUE #1: Develop a microservice that sends tasks to a Spark cluster in response to Kafka messages
The microservice has the following functionalities:
1. Accept a yml or json configuration file mapping a Kafka topic to a list of Spark scripts. E.g.:
   
   ```
   topic1:
       script1:
           [dependencies: list_of_dependencies_need_to_run_the_script]
       script2:
           [dependencies: list_of_dependencies_need_to_run_the_script]
   topic2:
       script3:
           [dependencies: list_of_dependencies_need_to_run_the_script]
   ```
   
   Note: Use the following library to transparently handle the configuration file: https://github.com/spf13/viper. Give a look at the [benchflow-client](https://github.com/benchflow/client) for an example on how to use Viper.
2. Wrap the [spark-submit](http://spark.apache.org/docs/latest/submitting-applications.html) tool.
3. Listen to the Kafka topics described in the configuration file. It starts different go routines listening for a topic each. For the interaction with Kafka we must use a protocol that ensure that we don't loose messages.
4. Submit Tasks to Spark, according to the configuration file. Use the [Builder pattern](https://en.wikipedia.org/wiki/Builder_pattern) to build the spark-submit commands. Example here: https://github.com/benchflow/client/tree/dev/src/cloud/benchflow/utils/docker. The script must be submitted with the cluster mode because we don't want to wait for the result.
5. Runs in a Docker container that obtain Spark scripts for the following two benchflow repositories: https://github.com/benchflow/analysers, https://github.com/benchflow/data-transformers and make them available for the submission to Spark.

### PR DESCRIPTION
**Pull request by Cerfoglg**
Developed task sender for issue #1

Closes #1

### CODE REVIEW

File: spark-task-sender/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,21 @@
+FROM golang
```
Comment (VincenzoFerme) [inline spark-task-sender/Dockerfile:1]: Needs to be rewritten using the correct FROM

File: spark-task-sender/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+RUN go install sender
+
+RUN echo "deb http://ppa.launchpad.net/webupd8team/java/ubuntu trusty main" > /etc/apt/sources.list.d/webupd8team-java.list
+RUN echo "deb-src http://ppa.launchpad.net/webupd8team/java/ubuntu trusty main" >> /etc/apt/sources.list.d/webupd8team-java.list
+RUN apt-key adv --keyserver keyserver.ubuntu.com --recv-keys EEA14886
+RUN apt-get -y update
+RUN /bin/echo debconf shared/accepted-oracle-license-v1-1 select true | /usr/bin/debconf-set-selections
+RUN DEBIAN_FRONTEND=noninteractive apt-get -y install oracle-java7-installer oracle-java7-set-default curl
+
+RUN curl -s http://d3kbcqa49mib13.cloudfront.net/spark-1.5.1-bin-hadoop2.6.tgz | tar -xz -C /usr/local/
```
Comment (VincenzoFerme) [inline spark-task-sender/Dockerfile:19]: Put the version as ENV variables. Example on the following link: https://github.com/benchflow/docker-images/blob/master/rancher-compose/Dockerfile#L6

File: spark-task-sender/src/sender/sender.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	}
+
+func constructSubmitCommand(ss SparkSubmit) exec.Cmd {
+	cmd := exec.Command(sparkHome+"/bin/spark-submit", "--master", sparkMaster, "--files", ss.MappingConfiguration, "--packages", "TargetHolding:pyspark-cassandra:0.1.5", ss.Script, sparkMaster, cassandraHost, ss.fileLocation)
+	return *cmd
+	}
+
+func consumeFromTopic(name string) {
+	go func() {
+		config := sarama.NewConfig()
```
Comment (VincenzoFerme) [inline spark-task-sender/src/sender/sender.go:63]: Move Kafka connection in a dedicated method

File: spark-task-sender/src/sender/sender.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+type SparkSubmit struct {
+	Dependencies []string
+	Script string
+	MappingConfiguration string
+	fileLocation string
+	}
+
+func constructSubmitCommand(ss SparkSubmit) exec.Cmd {
+	cmd := exec.Command(sparkHome+"/bin/spark-submit", "--master", sparkMaster, "--files", ss.MappingConfiguration, "--packages", "TargetHolding:pyspark-cassandra:0.1.5", ss.Script, sparkMaster, cassandraHost, ss.fileLocation)
```
Comment (VincenzoFerme) [inline spark-task-sender/src/sender/sender.go]: Use the builder pattern to build the command

File: spark-task-sender/src/sender/sender.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+	err = json.Unmarshal(dat, &c)
+	if err != nil {
+			panic(err)
+			}
+	
+	for _, sett := range c.Settings {
+		consumeFromTopic(sett.Topic)
+		}
+	
+	for true{}
```
Comment (VincenzoFerme) [inline spark-task-sender/src/sender/sender.go]: What is the purpose of this code?

File: spark-task-sender/Dockerfile (hunk 1 above)
Comment (Cerfoglg) [inline spark-task-sender/Dockerfile:1]: What is the correct FROM in this case? A specific version of GOLANG?

File: spark-task-sender/Dockerfile (hunk 2 above)
Comment (Cerfoglg) [inline spark-task-sender/Dockerfile:19]: Do you mean the version of GOLANG?

File: spark-task-sender/src/sender/sender.go (hunk 5 above)
Comment (Cerfoglg) [inline spark-task-sender/src/sender/sender.go]: I was going to use that loop to write some logs. For now it's empty, and it's there so the program doesn't quit while the other routines are running.

File: spark-task-sender/Dockerfile (hunk 1 above)
Comment (VincenzoFerme) [inline spark-task-sender/Dockerfile:1]: I'll take of it after merging the code. Basically we need to compile the go code and completely avoid the dependency from Go.

File: spark-task-sender/Dockerfile (hunk 2 above)
Comment (VincenzoFerme) [inline spark-task-sender/Dockerfile:19]: No, the Spark version: 1.5.1

File: spark-task-sender/src/sender/sender.go (hunk 5 above)
Comment (VincenzoFerme) [inline spark-task-sender/src/sender/sender.go]: Give a look at this approach: http://adampresley.com/2015/02/16/waiting-for-goroutines-to-finish-running-before-exiting.html. The same can be also used in other cases you have for monitors and collectors when you need to stop a goroutine that runs indefinitely.

**Comment by Cerfoglg** [comment]
Fixed builder pattern for spark command build

File: spark-task-sender/Dockerfile (hunk 2 above)
Comment (VincenzoFerme) [inline spark-task-sender/Dockerfile:19]: Now that you have the version as an environment variable, you should use it in place of "1.5.1" in the url. It is better if you move the ENV definition at the top of the file.

**Comment by Cerfoglg** [comment]
Added waitGroup to wait on goroutine completion. At the moment I haven't added the option to stop the goroutines, for now they keep on listening on the kafka topics. More graceful termination and error handling will be added in the future.

**Comment by VincenzoFerme** [comment]
@Cerfoglg let me know when all the issues in this pull request are fixed, so that I can work on it.

**Pull request merged**